File: src/main/java/min/queue/DesignCircularQueue.java
package min.queue;

public class DesignCircularQueue {

    // https://leetcode.com/explore/learn/card/queue-stack/228/first-in-first-out-data-structure/1396/
    static class MyCircularQueue {
        private int[] elements;
        private int front;
        private int rear;

        public MyCircularQueue(int k) {
            elements = new int[k];
            front = -1;
            rear = -1;
        }

        public boolean enQueue(int value) {
            if (isFull())
                return false;

            if (front == -1)
                front++;

            rear = (rear + 1) % elements.length;
            elements[rear] = value;
            return true;
        }

        public boolean deQueue() {
            if (isEmpty())
                return false;

            if (front == rear) {
                front = -1;
                rear = -1;
            } else {
                front = (front + 1) % elements.length;
            }

            return true;
        }

        public int Front() {
            return isEmpty() ? -1 : elements[front];
        }

        public int Rear() {
            return isEmpty() ? -1 : elements[rear];
        }

        public boolean isEmpty() {
            return front == -1 && front == rear;
        }

        public boolean isFull() {
            return (rear + 1) % elements.length == front;
        }
    }
}

File: src/main/java/thkim/queue/DesignCircularQueue.kt
package thkim.queue

class MyCircularQueue(k: Int) {

    private val buffer: Array<Int?> = arrayOfNulls(k)

    private var frontPos = 0 // 현재 값을 가리키고 있음
    private var rearPos = 0 // 다음 값을 가리키고 있음
    private var len = 0

    fun enQueue(value: Int): Boolean {
        if (buffer[rearPos] != null) return false

        buffer[rearPos] = value

        rearPos = (rearPos + 1) % buffer.size

        len++

        return true
    }

    fun deQueue(): Boolean {
        if (buffer[frontPos] == null || buffer.isEmpty()) return false

        buffer[frontPos] = null

        frontPos = (frontPos + 1) % buffer.size

        len--

        return true
    }

    fun Front(): Int = buffer[frontPos] ?: -1

    fun Rear(): Int = buffer[curRearPos()] ?: -1

    fun isEmpty(): Boolean = len == 0

    fun isFull(): Boolean = len == buffer.size

    fun curRearPos(): Int = if (rearPos - 1 < 0) buffer.size - 1 else rearPos - 1
}